sync/src/types/mod.rs

```diff
@@ -1718,29 +1752,31 @@ impl SyncState {
     }
 
     pub fn add_ask_for_txs(&self, peer_index: PeerIndex, tx_hashes: Vec<Byte32>) {
-        let mut unknown_tx_hashes = self.unknown_tx_hashes.lock();
-        if unknown_tx_hashes.len() >= MAX_UNKNOWN_TX_HASHES_SIZE {
-            return;
-        }
-
-        for tx_hash in tx_hashes
-            .into_iter()
-            .take(MAX_UNKNOWN_TX_HASHES_SIZE - unknown_tx_hashes.len())
-        {
-            match unknown_tx_hashes.entry(tx_hash) {
-                keyed_priority_queue::Entry::Occupied(entry) => {
-                    let mut priority = entry.get_priority().clone();
-                    priority.push_peer(peer_index);
-                    entry.set_priority(priority);
-                }
-                keyed_priority_queue::Entry::Vacant(entry) => {
-                    entry.set_priority(UnknownTxHashPriority {
-                        request_time: Instant::now(),
-                        peers: vec![peer_index],
-                        requested: false,
-                    })
+        let peer_unknown_count = self.peers().get_unknown_tx_hashes_count(peer_index);
+        let limit = MAX_UNKNOWN_TX_HASHES_SIZE.saturating_sub(peer_unknown_count);
+        if limit > 0 {
```

Comment:
Introducing `unknown_tx_hashes_count` is a strict way to enforce the number of requests to peers. And we have to carefully keep the consistency between `unknown_tx_hashes` and `unknown_tx_hashes_count`.


I suggest an alternative way to mitigate the problem without introducing a new field.

```
-        if unknown_tx_hashes.len() >= MAX_UNKNOWN_TX_HASHES_SIZE {
+        if unknown_tx_hashes.len() >= self.peers().len() * MAX_UNKNOWN_TX_HASHES_SIZE {
+            let mut peer_unknown_counter = HashMap::new();
+            for (_hash, priority) in unknown_tx_hashes.iter() {
+                for peer in priority.peers.iter() {
+                    peer_unknown_counter
+                        .entry(peer)
+                        .and_modify(|e| *e += 1)
+                        .or_insert(1);
+                }
+            }
+            for (peer, count) in peer_unknown_counter {
+                if count > MAX_UNKNOWN_TX_HASHES_SIZE {
+                    // punish this peer
+                }
+            }
+
```